#67284 Dashboard module + extending (#373)
valtimo-platform/valtimo-documentation: extending-valtimo/dashboard/custom-data-source.md

# Creating custom data sources

## Back-end

> The dependency `com.ritense.valtimo:dashboard` can be used to develop a back-end for a data source.

Reviewer: ```suggestion
The dependency `com.ritense.valtimo:dashboard` is required to develop a data source in the Valtimo back-end.
```

### Data source specification

> First, a data source specification using the `@WidgetDataSource` annotation needs to be created. Below is an example

Reviewer: ```suggestion
Data source specifications can be created using the `@WidgetDataSource` on a method in a Spring bean. Below is an example
```

Reviewer: I think the first here is a bit odd, this suggestion avoids that.

data source:

```kotlin
@Component
class RandomNumberWidgetDataSource {

    @WidgetDataSource(key = "randomNumber", title = "Random number")
    fun randomNumberDataSource(properties: RandomNumberDataSourceProperties): RandomNumberResult {
        return RandomNumberResult(Random.nextInt(properties.from, properties.to))
    }
}
```

The `@Component` is necessary to register the new data source as a Spring bean.

The data source key `randomNumber` must be unique and is used the identify the data source. The same key must also be
used in the front end data source.

The function can be named anything.

> The function can have either 0 or 1 argument. The single argument must be an object that contains the same fields that

Reviewer: I would add i little bit of explanation here. something like this:

If the datasource requires any configuration this will be passed in through the arguments of the function. All properties should be contained in a single object that is used as the argument for this function. The configuration from the front-end will be deserialized to this type, so the json structure should match that of the object. If no configuration is required the argument can be left out.
